File: Concurrency/Java/print-foobar-alternately.java
class FooBar {
    private int n;
    private Semaphore waitFirst;
    private Semaphore waitSecond;
    
    public FooBar(int n) {
        this.n = n;
        this.waitFirst = new Semaphore(0);
        this.waitSecond = new Semaphore(0);
    }

    public void foo(Runnable printFoo) throws InterruptedException {
        for (int i = 0; i < n; i++) {
            // printFoo.run() outputs "foo". Do not change or remove this line.
            printFoo.run();
            waitFirst.release();
            waitSecond.acquire();
        }
    }

    public void bar(Runnable printBar) throws InterruptedException {
        for (int i = 0; i < n; i++) {
          waitFirst.acquire();
          // printBar.run() outputs "bar". Do not change or remove this line.
          printBar.run();
          waitSecond.release();
        }
    }
}

File: Concurrency/C#/print-foobar-alternately.cs
using System.Threading;  
public class FooBar {
    private int n;
    private EventWaitHandle waitFirst;
    private EventWaitHandle waitSecond;
    
    public FooBar(int n) {
        this.n = n;
        this.waitFirst = new AutoResetEvent(initialState: false);
        this.waitSecond = new AutoResetEvent(initialState: false);
    }

    public void Foo(Action printFoo) {
        for (int i = 0; i < n; i++) {
        	printFoo();
            waitFirst.Set();
            waitSecond.WaitOne();
        }
    }

    public void Bar(Action printBar) {
        for (int i = 0; i < n; i++) {
            waitFirst.WaitOne();
            printBar();
            waitSecond.Set();
        }
    }
}
